### DISCUSSION TITLE
Slack Pricing and Plan Changes - Feedback Wanted

### DISCUSSION THREAD

**Discussion by myitcv** [cue-lang/cue#1839] [Q&A]
Hello all, 

Slack are due to make a number of [pricing and plan changes](https://slack.com/blog/news/pricing-and-plan-updates) on 1 Sep 2022. Currently, the [CUE community Slack](https://cuelang.slack.com) is on the free plan, which allows for retention of 10,000 messages: this gives the CUE community access to messages as far back as October 31 2020, equivalent to ~700 days. After September 1 2022, data retention will be a rolling 90 day window: on September 1 2022, this would revoke CUE community access to messages from before roughly 3 June 2022. This is therefore a significant change.

###  Goals 

Having a chat or messaging component like Slack for the CUE project is important for building the community and fostering more casual conversations ([see previous discussion](https://github.com/cue-lang/cue/discussions/1160)). It has a lower barrier for engagement and could even be considered the frontline of support. We therefore believe the CUE community needs a place to chat. 

 
Furthermore, we believe that retaining message history is an important feature for the CUE project. It provides history and continuity, and can be used as knowledge sharing and context building for the community. If this belief is shared by the wider community, we’ll need to decide on a course of action, because staying on the free plan does not satisfy this requirement. We need your help and feedback to reach a decision on what to do. Given the September 1 changeover date, we need to move fast!

### Criteria 

We have drafted a set of criteria against which we will evaluate options (see next section for those options):

* Historical message retention.
* Affordable.
* Ability to scale.
* Easy to maintain.
* Easy to moderate.
* Pre-existing familiarity:
    * UI/UX.
    * Use in related projects.
* Ability to integrate apps/plugins/bots
* Automated invite/add workflows.
* Open source a bonus.
* Migration/switching costs.
* Setup time for the new system.

Is there anything you would add/remove/change in this list?

_A note on affordability: this is hard to control with platforms. We can’t predict whether tools will pull the plug on free offerings or switch to a per-user pricing. So by definition, affordability is a point-in-time assessment._

### Options: 

* **Pro Slack at Full Price:** Paying per-user for Pro will be expensive for the CUE Project. At 918 users and growing, the pro user account would cost $8032.50 _per month_. And it will only get more costly with each new user. 
* **Pro Slack at Discount Price:** Currently, there are no such Open Source plans available for Slack. We are in touch with various people to see if Slack has received discount requests for the Pro Plan from other OSS projects, and what options OSS projects like CUE have. _NOTE:_ This may take some time and go beyond the September 1 deadline, so we are also looking into whether we can keep message retention while this is worked out. 
* **Remain on the Free Slack Plan:** This means we lose the ability to retain messages longer than 90 days. The other change in the free plan deals with storage limits that have historically not affected the CUE project, but this could be an issue once the project grows. 
* **Migrate to a Slack Alternative for CUE messaging:** we are currently exploring two free (or affordable) Slack alternatives: [Discord](https://discord.com) and [Matrix](https://matrix.org/) (see [Element](https://element.io/)). We will share more information about this exploration in this discussion as we learn more.

### Feedback Wanted

* How important is preserving messaging history to you?
* Are there decision tradeoffs that need mentioning? Benefits, drawbacks and switching costs that need to be considered?
* Are we missing any options which meet the criteria? 
* Are any criteria to be added, removed, changed? 
* Any challenges and risks to switching from Slack to an alternative like Discord or Matrix?
* Are there any tools to do user-level message export outside of Slack? (see below)
* Any other comments?

### Timeline

* **August 15:** Target deadline for feedback in this discussion (CUE maintainers to help collate feedback reported on Slack).
* **August 15-25:** Research and trial Slack alternatives. Updates posted to this discussion.
* **August 25:** CUE maintainers to publish analysis on options.
* **August 29:** Final reminder to the CUE community about the risk of data loss (see “Be prepared just in case” below).
* **September 1:** CUE maintainers to confirm final decision 

### Be prepared just in case

Since we don’t know if the decision to remain on Slack will be resolved by September 1st, you should take some time to copy any personal useful bits of information kept in Slack and transfer it somewhere more stable. Also **seeking community input here for user-level tools to transfer user messages (including threaded messages) off of Slack.** We will still be able to access Slack beyond September 1st, but unless we opt to pay for a Pro Plan (with a discount) we will only have 90 days of history. 

Thank you!

**Comment by wonderflow** [comment]
From my perspective, preserving history message is not that important. Slack is usually a place for Q&A, we can intentionally migrate important discussion to be github issue or github discussion.

**Reply by wonderflow** [reply]
And use slack alternatives May lose community users. Most community members are used to slack.

**Comment by marco-m** [comment]
> Migrate to a Slack Alternative for CUE messaging: we are currently exploring two free (or affordable) Slack alternatives: [Discord](https://discord.com/) and [Matrix](https://matrix.org/) (see [Element](https://element.io/)). We will share more information about this exploration in this discussion as we learn more.

I suggest considering also https://zulip.com/. Open source, free plan for open source projects (https://zulip.com/for/open-source/), or host yourself. API: https://chat.zulip.org/api/rest. Example of community: https://asciidoctor.zulipchat.com/

The reason why I am impressed with Zulip, besides it being open source, is that it is a mix of IM and forum, where it is easy to find information and the UI prods you to use proper threading, instead of just dumping a message in a channel.

**Reply by mvdan** [reply]
A tool that is a mix of IM and a forum such as Zulip, and which enforces the use of threads, might overlap a bit too much with our other non-chat forums like GitHub Discussions. It's all on a spectrum, but when I think "instant chat", I'm thinking of being able to quickly send a message without starting a thread, seeing when others are typing, seeing who is online, etc.

**Comment by mhsemcheski** [comment]
I think having a presence on Slack is extremely useful.  While other forums may be cheaper or have a similar set of features... I am in Slack, I have a number of other relevant workspaces already open.  That's not going to change, but I would be resistant to adding another messaging platform to my daily-configuration.  I say this not because its crucial that you keep me in the community, but because I think my reasoning here is not unique and lots of others wouldn't migrate to Dischord or Matrix or X.

There are a few times when I could see the value of looking something up that was older than 90 days, but its definitely not worth onboarding another messaging client.  I like the use of github for discussions and issues that may have more persistence.

My $0.02.

**Comment by august-feng** [comment]
I've done a lot of my initial CUE learning by scavenging the chat history like it is a knowledge base, so I'm grateful for the longer history retention the channel has had so far.

It's been a while since I've gone digging, but still I owe a lot of my learning to the chat history. As CUE gets adopted more, I imagine there will be a much better pool of resources to learn from for newcomers that are like me.

It's also been a pleasant experience of having the CUE community chat in the same chat application I use at work.

**Reply by mvdan** [reply]
I have similarly found the long chat history useful. A couple of times I have asked about a potential feature and the answer has been along the lines of: here is a link to the thread on github, but there was a very long and interesting chat discussion on Slack over a year ago. And it's understandable: if two or three people want to quickly bounce ideas, it's easier to use instant chat than a forum-like platform such as a GitHub issue or discussion.

**Comment by b4nst** [comment]
Same opinion as @wonderflow. I see the slack workspace as a day to day help/announcemens platform. For deeper subjects, GitHub discussions or issues are easier to track.

**Comment by qjcg** [comment]
I'm personally a fan of Matrix (and the Element client), and would be really happy to see this project use this open standard protocol and open-source project going forward.

There's also great potential for "meeting people where they are", that is, on whatever other chat service they may be using, via [Matrix Bridges](https://matrix.org/bridges/) (for example, the [matrix <-> slack bridge](https://matrix.org/docs/projects/bridge/matrix-appservice-slack)).

**Comment by cedricgc** [comment]
My order of preference:
1) Keep Slack with discount: The message history is important and search has long tail benefit (Also, save some of the early threads with Marcel talking about key decision points)
2) Discord: Wide user base, super flexible, coordinating voice calls are easier, can scale and keep messages
3) Matrix

**Comment by verdverm** [comment]
Generally
- the goals are sound, though I'm not sure a chat platform is the best to retain FAQs, more on this later
- I use Slack today as my primary messaging platform, I will not keep another regularly open. Slack definitely has a much more professional vibe than the others.
- It seems unreasonable for both OSS and for-profit organizations to pay so much to provide chat based support. They have some sort of public channel concept, though I don't know the details. Keep pushing to find lower priced options.
- If something like Matrix provides the team with a pane across several chat platforms, that might seem like a win.
- What are the Gopher slack and other major projects doing about this?
- I think the $XX,000 that would be spent on Slack could be put to better use, the user count could easily double in a year.

On retention
- we previously tried to move questions to GH discussions, though I'm not sure how much this improved the situation. It still seems like questions are asked and answered repeatedly. GH search had been awful, can't do partial/prefix string matching, which might be why.
- ultimately, I think we all agree documentation is the best way to address this sustainably
- to this end, Slack and GHD can both be the frontend to support, and more generally meeting users wherever they are is considered a good idea. This might mean the CUE team has many more platforms to monitor, which is something they have been trying to avoid. Are there any services or tools that can create a single dashboard to them?
- how can we reduce the effort to capture the most interesting content, those back and forth discussions? I manually copied some out, they are in some markdown files somewhere. Perhaps a slack bot & github repo for housing them could largely automate this? What about other platforms, both chat and forum based (Stack Overflow)?

My thoughts are to keep Slack, go with the free plan near-term, and hope that they walk back or provide some means that is acceptable for companies using their platform. Even if the CUE team no longer wants to officially support the CUE slack, it would be nice if someone else could maintain it on the free tier.

Another thought, long-term, if CUE is successful, there will be people talking about it in all sorts of places, and the CUE team will not be the source of frontline support. People and companies will start to build this expertise internally, much like Go looks today.

**Comment by Lyoness** [comment]
### Update

Thanks everyone for your feedback. Per the timeline above, we are updating today with an analysis of chat options open to the CUE project.

Based on analysis of your feedback, **it's likely the CUE Community will stay on Slack’s free plan**, with a rolling 90 day window of history. This tentative decision allows a final period for comments before 1 Sep 2022.

#### Rationale
Staying on Slack Free plan is the least disruptive path forward.  We’ve heard people’s concerns about having too many chat clients. People also want to benefit from chatting about CUE while at work, and keeping Slack is the best option for this for most. Message retention was important because it served as a knowledge base. The 90 day retention limit will force us as project maintainers to stop relying on chat as a knowledge base and improve our information porting processes to more appropriate places for the long-term benefit of all. 

#### Analysis of alternatives: 

**Why not paid Slack?** Please see the original post above. The per-user pricing is prohibitive, especially as the CUE project grows. If we do continue on the free plan, but find that a rolling 90 day window of history is not enough, we can revisit our options.
**Why not Discord or Zulip?** Switching to a new chat has benefits but also drawbacks.  We haven't seen strong community support for either of these; more people asked us not to force them to use  an entirely different tool, and/or to move off of Slack. Thus, we feel that the drawbacks of migrating off Slack outweigh any benefits of moving to Discord or Zulip. There are some feature requests that work better in these alternatives, and we’ve heard from larger OSS communities that recommended Discord. However, at this point in time, your feedback leads us to not pursue them. 
**Why not both Slack AND Discord?** At this point in CUE’s maturity as a project, it doesn’t make sense to have multiple chats. We can reassess this in future.
**Why not get rid of chat altogether and just use GH Discussions?** As discussed [before](https://github.com/cue-lang/cue/discussions/1160), chat is valuable for the community. The CUE open source project always intends to have a place for chat as one component of overall communication. 
**Why not an alternative or Open Source solution?**  We promised to evaluate alternatives, and to that end, we will run a trial with an open Source option . We’ll need to do this alongside using Slack’s free plan, so we can give the community time to evaluate without needing to make a hasty decision. Details to come in a separate post in this thread.
 
## Next steps: Take action before September 1, 2022

The tentative decision to stay on Slack’s free plan allows a final period for comments before 1 Sep 2022. After that date, messages older than 90 days for cuelang.slack.com will not be accessible. Take some time to go through Slack in the next week if you want to copy/paste messages to prepare for this.  Please continue to give feedback on whether the new free Slack policy change works (or doesn’t work) for the community at large.

Any final decisions made are only as good as the feedback that’s received, so please chime in if this adversely affects you.

Thanks,
-Carmen

**Comment by kishaningithub** [comment]
I would like to add a suggestion of a self hosted mattermost 

https://mattermost.com/

It has all the features of slack and is fully opensource
